### ISSUE TITLE
GraphQL query support

### ISSUE THREAD

**Issue by koraytaylan** [EmakinaTR/makina-api#18]
- [x] Choose the library to go with our current stack
- [x] Implement GraphQL query support for available endpoints
- [x] Provide some sample use cases in the postman collection file
- [x] Add unit tests for gql queries and mutations
~~Try [dataloader library](https://github.com/facebook/dataloader) to fetch entity relations.~~ Work on later in this issue.

**Comment by mmayadag** [comment]
[https://www.prisma.io/blog/relay-vs-apollo-comparing-graphql-clients-for-react-apps-b40af58c1534](https://www.prisma.io/blog/relay-vs-apollo-comparing-graphql-clients-for-react-apps-b40af58c1534
)

Firstly , We must choose what to use

There are a few alternatives for that
[Apollo Client](https://www.apollographql.com/docs/react/)
[Relay](https://facebook.github.io/relay/) 

If you need anything , I can help

**Comment by gokaygurcan** [comment]
Apollo is a no-brainer here. No need to add Relay into the mix. Apollo has way more support than Relay anyway.

**Comment by mmayadag** [comment]
Did you mean that ? 
![image](https://user-images.githubusercontent.com/33685760/51242335-c0ac1980-1990-11e9-9bc5-238fbc85def6.png)

**Comment by mmayadag** [comment]
Apollo is realy easy to use ,
There is a way to use postman but I recommend that software instead of postman 
https://altair.sirmuel.design/

**Comment by gokaygurcan** [comment]
https://en.wiktionary.org/wiki/no-brainer

**Comment by mmayadag** [comment]
> Graphql  = gql
https://graphql.org/
## Server Setup for Koa and Back-End
[Setting up Apollo Server with Koa](https://www.npmjs.com/package/apollo-server-koa)<<start here 
[Apollo server #Koa](https://github.com/apollographql/apollo-server)
[Graphql server getting started tutorial ](https://www.apollographql.com/docs/apollo-server/getting-started.html)

## Query and Playground tools 
[A beautiful feature-rich GraphQL Client for all platforms](https://altair.sirmuel.design/)**<<try**
or
[This is playground and desktop software for gql 
](https://github.com/prisma/graphql-playground)

---
This is api for api side , we will use same query files BE & FE.

There is only 1 endpoint which is
🚀 Server ready at http://localhost:4000/graphql

+ User Authentication 
+ etc 

We don't need postman & we don't need swagger , Don't think like many rest endpoint. 

Roadmap
1.  Crate services like ( micro services )
2. Bind your services with apollo-server for koa
3. Create gql and test it - make sure is it working - , << we will use those gql files at FE & BE & APP

Summarize , create your function and add it to graphql schema thats all for BE.

[_How to use gql with Postman_](https://www.codetraveler.io/2019/01/12/how-to-use-postman-with-a-graphql-api/) ( **don't use that** )

**Comment by ilhan-mstf** [comment]
There is an issue with how graphql works and how typeorm handles one-to-many and many-tp-many relations. For example; consider profile entity. This is a sample gql: 
```
{
  profile(id: 9) {
    id
    email
    place {
      id
      name
    }
  }
}
```
And it returns:
```
{
  "data": {
    "profile": {
      "id": "9",
      "email": "a@a.com",
      "place": null
    }
  }
}
```
As you can see that typeorm returns place as null since it only fetches profile. [In the documention](https://github.com/typeorm/typeorm/blob/master/docs/many-to-one-one-to-many-relations.md) they say that relations can be passed to function as a parameter (makes harder maintainability) or you can enable eager loading (which is not nice because of performance considerations). I don't like these solutions.  I'll look for other possible options, if you have suggestion please let me know.

**Comment by ilhan-mstf** [comment]
- [ ] Write unit tests for gql queries

**Comment by ilhan-mstf** [comment]
[This blog](https://blog.apollographql.com/optimizing-your-graphql-request-waterfalls-7c3f3360b051) post defines a way to fetch relations in a parallel way. It would be great if we can apply it in terms of performance. Furthermore it eliminates orm related stuffs. It a little bit complicates server side implementation however, it may worth to try out. This approach uses [dataloader library](https://github.com/facebook/dataloader). I'm going to try it but I guess it will take time.

**Comment by ilhan-mstf** [comment]
Closed by #32
